llvm/lib/Analysis/CaptureTracking.cpp

```diff
@@ -358,49 +363,54 @@ UseCaptureKind llvm::DetermineUseCaptureKind(
     // Volatile stores make the address observable.
     auto *ACXI = cast<AtomicCmpXchgInst>(I);
     if (U.getOperandNo() == 1 || U.getOperandNo() == 2 || ACXI->isVolatile())
-      return UseCaptureKind::MAY_CAPTURE;
-    return UseCaptureKind::NO_CAPTURE;
+      return CaptureInfo::otherOnly();
+    return CaptureInfo::none();
   }
   case Instruction::GetElementPtr:
     // AA does not support pointers of vectors, so GEP vector splats need to
     // be considered as captures.
     if (I->getType()->isVectorTy())
-      return UseCaptureKind::MAY_CAPTURE;
-    return UseCaptureKind::PASSTHROUGH;
+      return CaptureInfo::otherOnly();
+    return CaptureInfo::retOnly();
   case Instruction::BitCast:
   case Instruction::PHI:
   case Instruction::Select:
   case Instruction::AddrSpaceCast:
     // The original value is not captured via this if the new value isn't.
-    return UseCaptureKind::PASSTHROUGH;
+    return CaptureInfo::retOnly();
   case Instruction::ICmp: {
     unsigned Idx = U.getOperandNo();
     unsigned OtherIdx = 1 - Idx;
     if (auto *CPN = dyn_cast<ConstantPointerNull>(I->getOperand(OtherIdx))) {
```

Comment:
Unrelated to your PR, but isn't this somewhat predicate dependent. I.e `icmp slt ptr %X, null` also exposes the signbit.

Reply:
That's a good point. This code is also subtly incorrect for other reasons, but the signed predicate case is blatantly wrong. Not going to touch this now as this be removed altogether soon...

Reply:
Want me to throw up a quick PR to check that pred is equality? Or leave it all for later?

Reply:
I'd prefer to leave it for later.

Reply:
I ended up including the predicate check after all, as it's also needed for the new code, not just the old one.